src/core/hle/service/http/http_c.cpp

```diff
@@ -471,17 +724,115 @@ void HTTP_C::AddPostDataAscii(Kernel::HLERequestContext& ctx) {
+        IPC::RequestBuilder rb = rp.MakeBuilder(1, 2);
+        rb.Push(ResultCode(ErrCodes::InvalidRequestState, ErrorModule::HTTP,
+                           ErrorSummary::InvalidState, ErrorLevel::Permanent));
+        rb.PushMappedBuffer(buffer);
+        return;
+    }
+
+    http_context.post_data_raw.resize(buffer.GetSize());
+    buffer.Read(http_context.post_data_raw.data(), 0, buffer.GetSize());
+
+    IPC::RequestBuilder rb = rp.MakeBuilder(1, 2);
+    rb.Push(RESULT_SUCCESS);
+    rb.PushMappedBuffer(buffer);
+}
+
+void HTTP_C::GetResponseHeader(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx);
+
+    struct AsyncData {
+        HTTP_C* own;
+        u32 context_handle;
+        u32 name_len;
+        u32 value_max_len;
+        std::span<const u8> header_name;
+        Kernel::MappedBuffer* value_buffer;
+    };
+    std::shared_ptr<AsyncData> async_data = std::make_shared<AsyncData>();
+
+    async_data->own = this;
+    async_data->context_handle = rp.Pop<u32>();
+    async_data->name_len = rp.Pop<u32>();
+    async_data->value_max_len = rp.Pop<u32>();
+    async_data->header_name = std::span<const u8>(rp.PopStaticBuffer());
+    async_data->value_buffer = &rp.PopMappedBuffer();
+
+    if (!PerformStateChecks(ctx, rp, async_data->context_handle)) {
+        return;
+    }
+
+    ctx.RunAsync(
+        [async_data](Kernel::HLERequestContext& ctx) {
+            Context& http_context = async_data->own->GetContext(async_data->context_handle);
+            http_context.request_future.wait();
+            return 0;
+        },
+        [async_data](Kernel::HLERequestContext& ctx) {
+            std::string header_name_str(
+                reinterpret_cast<const char*>(async_data->header_name.data()),
+                async_data->name_len);
+            while (header_name_str.size() && header_name_str.back() == '\0') {
+                header_name_str.pop_back();
+            }
+
+            Context& http_context = async_data->own->GetContext(async_data->context_handle);
+
+            auto& headers = http_context.response.headers;
+            u32 copied_size = 0;
+
+            LOG_DEBUG(Service_HTTP, "header={}, max_len={}", header_name_str,
+                      async_data->value_buffer->GetSize());
+
+            auto header = headers.find(header_name_str);
+            if (header != headers.end()) {
+                std::string header_value = header->second;
+                copied_size = (u32)header_value.size();
+                if (header_value.size() + 1 > async_data->value_buffer->GetSize()) {
+                    header_value.resize(async_data->value_buffer->GetSize() - 1);
+                }
+                header_value.push_back('\0');
+                async_data->value_buffer->Write(header_value.data(), 0, header_value.size());
+            } else {
+                LOG_DEBUG(Service_HTTP, "header={} not found", header_name_str);
+                IPC::RequestBuilder rb(ctx, 0x001E, 1, 2);
```

Comment:
can we make this consistent?

```suggestion
                IPC::RequestBuilder rb = rp.MakeBuilder(1, 2);
```

same below

Reply:
`rp` is not available in the async context.